### PR TITLE
Fix IOperation for asynchronous using and foreach statements

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/30362

- add `IsAsynchronous` to `IUsingOperation` and `IForEachLoopOperation`
- add `IsAsynchronous` to `ForeachEnumeratorInfo`
- add missing awaits on invocations of `DisposeAsync()` and `MoveNextAsync()`
- correct references to `IDisposable` to be `IAsyncDisposable`

### CODE REVIEW

**Comment by jcouv** [comment]
@AlekseyTs I thought I'd have something more to do for CFG, but I think I've already handled it. Thanks

**Comment by jcouv** [comment]
@333fred for review. Thanks

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IForEachLoopStatement.cs [hunk 1]
```diff
@@ -3805,16 +3806,18 @@ static async System.Threading.Tasks.Task Main(System.Collections.Generic.IAsyncE
...
-IForEachLoopOperation (LoopKind.ForEach, Continue Label Id: 0, Exit Label Id: 1) (OperationKind.Loop, Type: null) (Syntax: 'await forea ... }')
+IForEachLoopOperation (LoopKind.ForEach, IsAsynchronous, Continue Label Id: 0, Exit Label Id: 1) (OperationKind.Loop, Type: null) (Syntax: 'await forea ... }')
   Locals: Local_1: System.String value
   LoopControlVariable: 
     IVariableDeclaratorOperation (Symbol: System.String value) (OperationKind.VariableDeclarator, Type: null) (Syntax: 'string')
       Initializer: 
         null
   Collection: 
-    IParameterReferenceOperation: pets (OperationKind.ParameterReference, Type: System.Collections.Generic.IAsyncEnumerable<System.String>) (Syntax: 'pets')
+    IConversionOperation (TryCast: False, Unchecked) (OperationKind.Conversion, Type: System.Collections.Generic.IAsyncEnumerable<System.String>, IsImplicit) (Syntax: 'pets')
```
Comment (333fred) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IForEachLoopStatement.cs]: Do you know why this conversion is introduced?

**Comment by 333fred** [comment]
Please add some tests with the IAsyncX interfaces missing in the compilation for both IOp and CFG generation.

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs [hunk 2]
```diff
@@ -600,6 +600,10 @@ private void LogLoopStatementHeader(ILoopOperation operation, bool? isChecked =
             var propertyStringBuilder = new StringBuilder();
             propertyStringBuilder.Append(" (");
             propertyStringBuilder.Append($"{nameof(LoopKind)}.{operation.LoopKind}");
+            if (operation is IForEachLoopOperation foreachLoop && foreachLoop.IsAsynchronous)
```
Comment (333fred) [inline src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: >foreachLoop && foreachLoop.IsAsynchronous [](start = 51, length = 41)

Nit: could use recursive pattern :)

**Comment by 333fred** [comment]
Done review pass (commit 4)

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IForEachLoopStatement.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IForEachLoopStatement.cs]: Not sure. The logic in BindForEachPartsWorker always wraps the collection expression with a BoundConversion (even when the conversion is Identity).
The conversion is there in other IOperation tests for regular foreach loops (`ForEachFlow_06`) so this seems correct.

Let me know if this bugs you. I can investigate further.

---
In reply to: [316309905](https://github.com/dotnet/roslyn/pull/37963#discussion_r316309905) [](ancestors = 316309905)

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for another review.
Thanks @333fred!

File: src/Compilers/Core/Portable/Generated/Operations.Generated.cs [hunk 3]
```diff
@@ -431,6 +436,11 @@ public interface IUsingOperation : IOperation
         /// Locals declared within the <see cref="Resources" /> with scope spanning across this entire <see cref="IUsingOperation" />.
         /// </summary>
         ImmutableArray<ILocalSymbol> Locals { get; }
+        /// <summary>
+        /// Whether this for using is asynchronous.
```
Comment (cston) [inline src/Compilers/Core/Portable/Generated/Operations.Generated.cs]: >for [](start = 25, length = 3)

Remove?

File: src/Compilers/Core/Portable/Operations/OperationInterfaces.xml [hunk 4]
```diff
@@ -465,6 +473,14 @@
         </summary>
       </Comments>
     </Property>
+    <Property Name="IsAsynchronous" Type="bool">
+      <Comments>
+        <summary>
+          Whether this for using is asynchronous.
```
Comment (cston) [inline src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: >for [](start = 23, length = 3)

Remove?

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 5]
```diff
@@ -4110,9 +4122,15 @@ public override IOperation VisitForEachLoop(IForEachLoopOperation operation, int
 
                 LeaveRegion();
 
+                bool isAsynchronous = info.IsAsynchronous;
+                var iDisposable = isAsynchronous
+                    ? _compilation.GetTypeByMetadataName(WellKnownTypes.GetMetadataName(WellKnownType.System_IAsyncDisposable))
```
Comment (cston) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >_compilation.GetTypeByMetadataName(WellKnownTypes.GetMetadataName(WellKnownType.System_IAsyncDisposable) [](start = 22, length = 104)

Should there be a `CommonGetWellKnownType()` method?

File: src/Compilers/CSharp/Portable/Binder/ForEachEnumeratorInfo.cs [hunk 6]
```diff
@@ -30,6 +30,8 @@ internal sealed class ForEachEnumeratorInfo
         // When async and needs disposal, this stores the information to await the DisposeAsync() invocation
         public AwaitableInfo DisposeAwaitableInfo;
 
+        public readonly bool IsAsync;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/ForEachEnumeratorInfo.cs]: >bool [](start = 24, length = 4)

Minor point: Perhaps move next to `NeedsDisposal` to reduce size of instance.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs (hunk 5 above)
Comment (cston) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Should use new `CommonGetWellKnownType()` method.

---
In reply to: [316817457](https://github.com/dotnet/roslyn/pull/37963#discussion_r316817457) [](ancestors = 316817457)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 7]
```diff
@@ -3841,7 +3852,8 @@ bool isNotNullableValueType(ITypeSymbol type)
 
         private IOperation ConvertToIDisposable(IOperation operand, ITypeSymbol iDisposable, bool isTryCast = false)
         {
-            Debug.Assert(iDisposable.SpecialType == SpecialType.System_IDisposable);
+            Debug.Assert(iDisposable.SpecialType == SpecialType.System_IDisposable ||
+                iDisposable.Equals( _compilation.GetTypeByMetadataName(WellKnownTypes.GetMetadataName(WellKnownType.System_IAsyncDisposable))));
```
Comment (cston) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >GetTypeByMetadataName [](start = 49, length = 21)

`CommonGetWellKnownType()`

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs (hunk 5 above)
Comment (jcouv) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Oops. Thanks

---
In reply to: [316915353](https://github.com/dotnet/roslyn/pull/37963#discussion_r316915353) [](ancestors = 316915353,316817457)

**Pull request merged**